Make user field for oauth subject configurable

ThoWagen opened the pull request:
Currently the subject in the tokens created for OAuth2 is always the id of the user table. But this id does not necessarily have any meaning for other services. This PR allows to configure the used user field for the subject.

ThoWagen committed:
Make user field for oauth subject configurable

demiankatz commented on module/VuFind/src/VuFind/Db/Table/User.php:
```diff
@@ -104,6 +104,19 @@ public function createRowForUsername($username)
+     * @param string $fieldName  Field name
+     * @param string $fieldValue Field value
+     *
+     * @return UserRow
```
Should this return `?UserRow` in case no match is found?

ThoWagen replied:
Yes, i also changed it for the other getter methods.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/User.php:
```diff
@@ -104,6 +104,19 @@ public function createRowForUsername($username)
+     *
+     * @return UserRow
+     */
+    public function getByField($fieldName, $fieldValue)
```
This change also comes at a somewhat unfortunate time, as I'm currently working on introducing an interface that uses standard getters and setters for all user properties, in order to better abstract away from the underlying database. Some of these changes kind of cut against that. But I suppose, worst case scenario, after the refactoring is done, we could introduce a map of field name to getter method in order to keep this working.

ThoWagen replied:
Since we only need this for id, username and cat_it now, I changed this so that the getter methods are used.

demiankatz replied:
Maybe it would be better to move this logic into the UserService. I don't think there's an advantage to adding a new method to the Table class, since the plan is to eliminate the Table class entirely as soon as we can... and this logic could easily be moved into UserService::getUserByField(). Or maybe it's safer still to just add UserService::getUserByUsername() and UserService::getUserByCatalogId() methods and move this switch statement into the calling code. That way, we don't introduce a general-purpose method that may prove difficult to port cleanly, but I don't think it increases the complexity of this PR significantly.

demiankatz reviewed (commented):
Thanks, @ThoWagen -- I've left a couple of very general thoughts, and requested a review from @EreMaijala, who I think is significantly more familiar with the OAuth2 code than I am.

EreMaijala reviewed (changes requested):
As far as I can see there are the following issues:

1. Only id, username and cat_id are guaranteed to be unique. You cannot use any other field without the risk of accessing wrong user.
2. Only a few columns in the user table have keys. Retrieving a user by any field not currently indexed is a relatively heavy operation particularly when the user table is large.
3. The subject in OIDC is not suposed to have a meaning for external services. It also must be non re-assignable. This is from the spec:

> Subject Identifier. Locally unique and never reassigned identifier within the Issuer for the End-User, which is intended to be consumed by the Client, e.g., 24400320 or AItOawmwtWwcT0k51BayewNvutrJUqsvl6qs7A4. It MUST NOT exceed 255 ASCII characters in length. The sub value is a case-sensitive string.

As a conclusion I think that the subject identifier should NOT be made configurable. Any other fields that have more meaningful content for the client can be carried in other claims.

ThoWagen commented:
@EreMaijala 

Regarding 1. and 2.: I see the problem that not every field is suitable for being the `sub`. I did not restrict the fields that can be used because there might be custom setups with other user table configurations. But we should probably only allow id, username and cat_it by default and also write a warning about changing the field. 

Regarding 3. and your conclusion: If I interpret it correctly  "which is intended to be consumed by the client" means that the OICD client actually needs to know how to interpret the subject. And this should also work for OAuth2 only. The `sub` is supposed to be an identifier that is also known by the resource server. ~~See https://datatracker.ietf.org/doc/html/rfc9068#section-2.2 and https://www.rfc-editor.org/rfc/rfc7519#section-4.1.2.~~ In our setup the username is an unique identifier that is used by all other services of the libraries and universities.

@demiankatz I have seen you comments and will come back to them if Ere and I sorted out if and how to move forward with this.

EreMaijala commented:
@ThoWagen I didn't mean to imply that subject isn't used by the client but that it's an opaque string that uniquely identifies the subject within the identity provider. See also [section 6 in RFC 9068](https://datatracker.ietf.org/doc/html/rfc9068#PrivacyConsiderations).

EreMaijala commented:
@ThoWagen Regarding point 1, an additional issue (that may or may not be a problem in a specific case) is that cat_id is only present with certain ILS logins.

ThoWagen commented:
@EreMaijala [section 6 in RFC 9068](https://datatracker.ietf.org/doc/html/rfc9068#PrivacyConsiderations) also states

> This profile mandates the presence of the "sub" claim in every JWT access token, making it possible for resource servers to rely on that information for correlating incoming requests with data stored locally for the authenticated principal

In our case the resource servers would use the username identify the principal.

ThoWagen commented:
> Regarding point 1, an additional issue (that may or may not be a problem in a specific case) is that cat_id is only present with certain ILS logins.

That's why we should place a warning in the configuration and explain what the requirements are for changing the used field.

ThoWagen commented:
If the risk that this leads to confusion or security problems is to high which I can not estimate we can implement in our custom module. But I thought that might be used by others as well.

EreMaijala commented:
> @EreMaijala [section 6 in RFC 9068](https://datatracker.ietf.org/doc/html/rfc9068#PrivacyConsiderations) also states
> 
> > This profile mandates the presence of the "sub" claim in every JWT access token, making it possible for resource servers to rely on that information for correlating incoming requests with data stored locally for the authenticated principal
> 
> In our case the resources owners would use the username identify the principal.

Yes, it's unique and can be used to correlate [1] the claim with the locally stored data, but the contents of the claim is not something the client should attempt to understand. Providing just a numeric ID allows correlation without revealing anything specific about the subject. If you need to connect the user with information gathered from somewhere else, you could map e.g. username to a username claim that the client would (in this specific case) use to gather further information about the user, but I'd still not support using username as sub.

[1] Well, only to the extent the server allows it to be. The RFC specifically mentions that the server may even provide a different sub for every single claim according to its privacy impact assessment.

ThoWagen commented:
>  but the contents of the claim is not something the client should attempt to understand

Yes, but the main reasoning for this is
>  inspect the content of the access token: the authorization server and the resource server might decide to change the token format at any time (for example, by switching from this profile to opaque tokens); hence, any logic in the client relying on the ability to read the access token content would break without recourse.

> The RFC specifically mentions that the server may even provide a different sub for every single claim according to its privacy impact assessment.

The reason for this is to prevent tracking of principal activities across multiple or even a single resource server. But this is also possible if we use the user's id.

> If you need to connect the user with information gathered from somewhere else, you could map e.g. username to a username claim that the client would (in this specific case) use to gather further information about the user, but I'd still not support using username as sub.

The resource server is supposed to work with the access token and not the id token that could contain the username as additional claim. The access token only specifies scopes.

The resource server could use the access_token to call the userInfo endpoint and retrieve the username that way. But I don't think that this is how it is supposed to work. JWTs are used for the reason resource servers don't have to validate the access tokens at the authorization server every time. And since most if not all of our resource servers would need the username scope the clients could get the username via the userInfo endpoint anyway.

We could make `sub` configurable for every client. That way we can set it to the username whenever the token is supposed to be used by a resource server that uses the username.

EreMaijala commented:
Well, I'm probably still missing a detail or two, but I suppose it'd be fine to make the id configurable as long as the comments are very clear about the consequences.

ThoWagen commented:
Unfortunately, I mixed up resource owner and resource server in the discussion. So I fixed it in my previous comments.
 
But my main argument for using the username as `sub` over an access_token with a username scope is that anyone who has such an access_token can acquire the username anyway. Everyone including the resource servers just need another request to the authorization server to do so.

But I can see that it would be possible to have tokens where the username is not necessary. So I will check if the configuration can be made per client. In that case we could also add the possibility of encrypting the username with a public key of a resource server. Then we would actually have unique `sub`s and that no one else is able to read.

EreMaijala commented:
> But my main argument for using the username as `sub` over an access_token with a username scope is that anyone who has such an access_token can acquire the username anyway. Everyone including the resource servers just need another request to the authorization server to do so.

No, they can't unless they've requested the scope and been allowed access to it [1]. If username is used for sub, the user should be properly informed as well.

[1] We don't currently have a way of choosing which claims to release when requested (and adding that would make it possible allow access while not providing enough information for the client to actually proceed properly, which could cause unexpected issues), but it would be possible to do.

ThoWagen commented:
> No, they can't unless they've requested the scope and been allowed access to it.

You are right. I just implicitly assumed that if a client requests an access token to access a resource server it always includes the username scope. But maybe that is a problem and the setup I have in mind can not be generalized.

Having client specific configurations for the identity repository is not really possible without reimplementing some parts of the league oauth2 server (i.e. getting the client id in the token action). So I would advise against it.

In my opinion using the username if it is an global user identifier as `sub` is not really problematic in our use case and does not contradict the oauth2 specifications. But I don't mind to close this PR and to either implement the changes in our custom module or just doing that extra request to acquire the username in our resource servers.

EreMaijala commented:
@ThoWagen Like I said before, I don't object making it configurable, as long as the comments indicate clearly what the consequences are. I'd also limit the fields to those that are guaranteed to be unique. And of course the user should also know what information about them is being revealed.

ThoWagen committed:
restrict allowed user identifier fields, informing users about used field and expanded config comment

ThoWagen commented:
@EreMaijala I updated the comment and restricted the configurable field to id, username and cat_id. For the username and cat_id an extra scope will be displayed to the user.

EreMaijala commented on languages/en.ini:
```diff
@@ -461,6 +461,7 @@ external_auth_scope_address = "Read your address"
 external_auth_scope_age = "Read your age"
 external_auth_scope_birthdate = "Read your date of birth"
 external_auth_scope_block_status = "Check if your account has blocks"
+external_auth_scope_cat_id = "Read your catalog user identifier"
```
I'd formulate this along the similar ones, like "Read your internal library user identifier".

ThoWagen replied:
Done

EreMaijala commented on module/VuFind/src/VuFind/Db/Table/User.php:
```diff
@@ -104,12 +104,35 @@ public function createRowForUsername($username)
+                return $this->getByUsername($fieldValue, false);
+            case 'cat_id':
+                return $this->getByCatalogId($fieldValue);
+            case 'email':
```
This won't be needed, right?

ThoWagen replied:
It is not needed but I left it since it might be used in the future. However, I can delete it if you prefer that.

EreMaijala replied:
I'd leave it out since there's no guarantee at all for it to be unique.

ThoWagen replied:
Being unique is not necessary for other possible use cases for this method. But off course if such a use case will occur it can be added again. So I will remove it for now.

ThoWagen replied:
Done

EreMaijala commented on module/VuFind/src/VuFind/OAuth2/Entity/UserEntity.php:
```diff
@@ -91,7 +91,14 @@ public function __construct(
+        $userIdentifier = $user->$userIdentifierField;
+        if ($userIdentifier == null) {
+            throw new \VuFind\Exception\BadConfig(
+                'User identifier is empty. The user identifier field needs to be required.'
```
Perhaps a bit more helpful error message like `"Required user identifier field $userIdentifierField empty for user {$user->id}`"?

ThoWagen replied:
Yes, that's better. Changed it

EreMaijala commented on module/VuFind/src/VuFind/OAuth2/Entity/UserEntity.php:
```diff
@@ -91,7 +91,14 @@ public function __construct(
-        $this->setIdentifier($user->id);
+        $userIdentifierField = $config['Server']['userIdentifierField'] ?? 'id';
+        $userIdentifier = $user->$userIdentifierField;
+        if ($userIdentifier == null) {
```
Maybe strict comparison `===`?

ThoWagen replied:
done

EreMaijala reviewed (changes requested):
Looks pretty good, just a couple of small suggestions.

ThoWagen committed:
Merge branch 'dev' into pull-request/configurable-oauth-subject

ThoWagen committed:
improved translations and error message

ThoWagen committed:
removed email from getByField

demiankatz reviewed (changes requested):
@ThoWagen, unfortunately, there's been a lot of activity in the OAuth code all at once, and there are now quite a few conflicts here. Hopefully they won't be too hard to resolve, but please let me know if I can be of any help. This is likely all related to #3499, if looking at the work there is helpful for reference during the conflict resolution.

Once this is sorted out, I'll be happy to merge the PR.

ThoWagen committed:
merged dev and refactored to match new implementation

ThoWagen committed:
Merge branch 'dev' into pull-request/configurable-oauth-subject

demiankatz commented on module/VuFind/src/VuFind/OAuth2/Repository/TokenRepositoryFactory.php:
```diff
@@ -44,7 +44,11 @@
  * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
  * @link     https://vufind.org Main Site
  */
+<<<<<<<< HEAD:module/VuFind/src/VuFind/OAuth2/Repository/TokenRepositoryFactory.php
```
@ThoWagen, this is a very fresh commit, so I realize you may still be working on it, but in case you haven't already seen this, there appear to be conflict markers committed to the PR at this point in time. :-)

ThoWagen replied:
Resolving the merge conflicts took some time so I forgot to finish the merge request and just committed the changes. So the merge got messed up unfortunately. But it should be fixed now finally. 

Sorry for the hick up.

demiankatz replied:
No problem, it will all go away when we "squash and merge" at the end of the process. :-)

ThoWagen committed:
allow null as return value in user service get methods

ThoWagen committed:
resolved commited merge conflict

demiankatz commented on module/VuFind/src/VuFind/Db/Row/User.php:
```diff
@@ -784,6 +784,18 @@ public function getLoginTokens(string $userId): array
+     *
+     * @return mixed
+     */
+    public function getField($fieldName)
```
Similarly, could we avoid adding getField() by just adding a switch to pick the appropriate getter based on the provided field name in the calling code? A long-term goal of the entity classes is to avoid exposing internal details like raw field names, so I think it would be best not to set this precedent here if there's a reasonably simple alternative.

demiankatz reviewed (changes requested):
Thanks, @ThoWagen, and sorry again for the inconvenience caused here. Just a couple of suggestions in the interest of keeping the database code as simple as possible during our current time of design transition. I'm certainly open to revisiting more general-purpose utility methods in the future, but until our long-term design is a bit more mature, I'd prefer to avoid introducing these things.

ThoWagen committed:
removed getField method

ThoWagen committed:
moved getByField implementation into UserService

ThoWagen commented:
Since we only allow three user fields as identifier I agree that the getter does not need to be generalized. I replaced `getField` by using a select statement as you suggested.

And I also moved the getByField logic into the `UserService`. Off course we could also get rid of it there and create a UserIdentityTrait or somthing like that. But I am not sure if that if that is an improvement.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/UserService.php:
```diff
@@ -63,10 +63,31 @@ public function __construct(protected User $userTable)
+    /**
+     * Retrieve a user object from the database based on the given field.
+     *
+     * @param string $fieldName  Field name
```
Should we add a list of supported field values to this comment?

ThoWagen replied:
done

demiankatz commented on module/VuFind/src/VuFind/Db/Service/UserService.php:
```diff
@@ -63,10 +63,31 @@ public function __construct(protected User $userTable)
+            case 'cat_id':
+                return $this->userTable->getByCatalogId($fieldValue);
+        }
+        return null;
```
Might it be better to throw an exception complaining about an unsupported field if we reach this point in the method?

ThoWagen replied:
Yes, that's a good idea. Done

demiankatz reviewed (changes requested):
Thanks, @ThoWagen -- a couple of small thoughts on the latest refactoring.

ThoWagen committed:
added allowed fields to doc comment and exception on invalid field

ThoWagen committed:
cs fix

demiankatz committed:
Merge branch 'dev' into pull-request/configurable-oauth-subject

demiankatz commented on module/VuFind/src/VuFind/Db/Service/UserServiceInterface.php:
```diff
@@ -47,7 +47,17 @@ interface UserServiceInterface extends DbServiceInterface
+     *
+     * @return ?UserEntityInterface
+     */
+    public function getUserByField($fieldName, $fieldValue);
```
Is it worth adding typehints to the method definition while we're at it? Might be good to do the same for getUserById, though that's perhaps out of scope for this PR.

ThoWagen replied:
I added the type hints.

I wonder what your general thought on type hints is. Do you want to have them in all classes/interfaces and we just have not done this as much previously or are they only desired in some cases as this one?

demiankatz replied:
@ThoWagen, I think it's generally desirable to add typehints to new code when possible. We haven't introduced them globally due to backward-compatibility concerns, but someday it would be nice to increase the level of type safety in the code, and if we add hints as we refactor, it will make that eventual job easier when we get there. :-)

demiankatz commented on module/VuFind/src/VuFind/OAuth2/Entity/UserEntity.php:
```diff
@@ -66,7 +66,26 @@ public function __construct(
+        }
+        if ($userIdentifier === null) {
+            throw new \VuFind\Exception\BadConfig(
+                "Required user identifier field $userIdentifierField empty for user {$user->id}"
```
Should we throw a different exception in the default case above? If an unsupported field is specified, that's not really an empty value error, so this message might be misleading.

ThoWagen replied:
I updated the message. Hopefully it is more helpful now.

demiankatz commented on module/VuFind/src/VuFind/OAuth2/Repository/IdentityRepository.php:
```diff
@@ -68,13 +68,14 @@ public function __construct(
      * Get a user entity by identifier.
      *
-     * @param int $identifier User ID
+     * @param int $identifier User Identifier
```
I don't think int is the right type here anymore, since some identifier types might be strings.

ThoWagen replied:
Fixed

demiankatz reviewed (changes requested):
@ThoWagen, I've resolved the conflicts I recently introduced to save you from having to do that yet again. I noticed a couple of small things while reviewing the code again; see below:

demiankatz commented:
I should also note that all tests are passing for me, but it wouldn't hurt for you to review my merging work in case I did anything wrong. :-)

ThoWagen committed:
Merge branch 'dev' into pull-request/configurable-oauth-subject

ThoWagen committed:
added and fixed type hints

ThoWagen commented:
I checked the current state and everything seems to work as expected.

demiankatz reviewed (approved):
Thanks, @ThoWagen -- looks good, and all tests are passing. Merging now!
